Detectors/MUON/MCH/DigitFiltering/src/DigitFilteringSpec.cxx

```diff
@@ -0,0 +1,169 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "MCHDigitFiltering/DigitFilteringSpec.h"
+
+#include "DataFormatsMCH/Digit.h"
+#include "DataFormatsMCH/ROFRecord.h"
+#include "Framework/ConfigParamRegistry.h"
+#include "Framework/DataSpecUtils.h"
+#include "Framework/Logger.h"
+#include "Framework/OutputSpec.h"
+#include "Framework/Task.h"
+#include "Framework/WorkflowSpec.h"
+#include "MCHBase/SanityCheck.h"
+#include "MCHDigitFiltering/DigitFilterParam.h"
+#include "SimulationDataFormat/MCCompLabel.h"
+#include "SimulationDataFormat/MCTruthContainer.h"
+#include <fmt/format.h>
+#include <iostream>
+#include <string>
+#include <vector>
+
+using namespace o2::dataformats;
+using namespace o2::framework;
+
+namespace o2::mch
+{
+class DigitFilteringTask
+{
+ public:
+  DigitFilteringTask(bool useMC) : mUseMC{useMC} {}
+
+  void init(InitContext& ic)
+  {
+    mSanityCheck = ic.options().get<bool>("sanity-check");
+    mMinADC = ic.options().get<int>("min-adc-value");
+  }
+
+  bool isGoodDigit(const Digit& digit) const
+  {
+    return digit.getADC() > mMinADC;
+  }
+
+  void run(ProcessingContext& pc)
+  {
+    // get input
+    auto iRofs = pc.inputs().get<gsl::span<ROFRecord>>("rofs");
+    auto iDigits = pc.inputs().get<gsl::span<Digit>>("digits");
+    auto iLabels = mUseMC ? pc.inputs().get<MCTruthContainer<MCCompLabel>*>("labels") : nullptr;
+
+    bool abort{false};
+
+    if (mSanityCheck) {
+      LOGP(info, "performing sanity checks");
+      auto error = sanityCheck(iRofs, iDigits);
+
+      if (!isOK(error)) {
+        if (error.nofOutOfBounds > 0) {
+          LOGP(error, asString(error));
+          LOGP(error, "in a TF with {} rofs and {} digits", iRofs.size(), iDigits.size());
+          abort = true;
+        }
+      }
+    }
+
+    // create the output messages
+    auto& oRofs = pc.outputs().make<std::vector<ROFRecord>>(OutputRef{"rofs"});
+    auto& oDigits = pc.outputs().make<std::vector<Digit>>(OutputRef{"digits"});
+    auto oLabels = mUseMC ? &pc.outputs().make<MCTruthContainer<MCCompLabel>>(OutputRef{"labels"}) : nullptr;
+
+    if (!abort) {
+      int cursor{0};
+      for (auto i = 0; i < iRofs.size(); i++) {
+        const ROFRecord& irof = iRofs[i];
+
+        const auto digits = iDigits.subspan(irof.getFirstIdx(), irof.getNEntries());
+
+        // filter the digits from the current ROF
+        for (auto i = 0; i < digits.size(); i++) {
```

Comment:
The variable 'i' masks the one of the 1st loop (this should issue a compilation warning?). I would use a range-based for loop for the rofs.

Reply:
agreed